core/benches/bench.rs

```diff
@@ -0,0 +1,732 @@
+                disk: None,
+                io_type,
+                io_pattern,
+                buf_size,
+                total_bytes,
+                concurrency,
+            }))
+        }
+    }
+
+    pub struct SimpleDiskBench {
+        name: String,
+        disk_type: DiskType,
+        disk: Option<Arc<dyn BenchDisk>>,
+        io_type: IoType,
+        io_pattern: IoPattern,
+        buf_size: usize,
+        total_bytes: usize,
+        concurrency: u32,
+    }
+
+    impl Bench for SimpleDiskBench {
+        fn name(&self) -> &str {
+            &self.name
+        }
+
+        fn total_bytes(&self) -> usize {
+            self.total_bytes
+        }
+
+        fn run(&self) -> Result<()> {
+            let io_type = self.io_type;
+            let io_pattern = self.io_pattern;
+            let buf_nblocks = self.buf_size / BLOCK_SIZE;
+            let total_nblock = self.total_bytes / BLOCK_SIZE;
+            let concurrency = self.concurrency;
+
+            let disk = self.disk.as_ref().unwrap();
+
+            let local_nblocks = total_nblock / (concurrency as usize);
+            let join_handles: Vec<JoinHandle<Result<()>>> = (0..concurrency)
+                .map(|i| {
+                    let disk = disk.clone();
+                    let local_pos = (i as BlockId) * local_nblocks;
+                    thread::spawn(move || match (io_type, io_pattern) {
+                        (IoType::Read, IoPattern::Seq) => {
+                            disk.read_seq(local_pos, local_nblocks, buf_nblocks)
+                        }
+                        (IoType::Write, IoPattern::Seq) => {
+                            disk.write_seq(local_pos, local_nblocks, buf_nblocks)
+                        }
+
+                        (IoType::Read, IoPattern::Rnd) => {
+                            disk.read_rnd(local_pos, local_nblocks, buf_nblocks)
+                        }
+                        (IoType::Write, IoPattern::Rnd) => {
+                            disk.write_rnd(local_pos, local_nblocks, buf_nblocks)
+                        }
+                    })
+                })
+                .collect();
+
+            let mut any_error = None;
+            for join_handle in join_handles {
+                let res = join_handle
+                    .join()
+                    .expect("couldn't join on the associated thread");
+                if let Err(e) = res {
+                    println!("benchmark task error: {:?}", &e);
+                    any_error = Some(e);
+                }
+            }
+            match any_error {
+                None => Ok(()),
+                Some(e) => Err(e),
+            }
+        }
+
+        fn prepare(&mut self) {
+            let disk = self.prepare_disk().unwrap();
+            let _ = self.disk.insert(disk);
+        }
+
+        fn display_ext(&self) {}
+    }
+
+    impl SimpleDiskBench {
+        fn prepare_disk(&self) -> Result<Arc<dyn BenchDisk>> {
```

Comment:
The functionality  of this method contains two parts, `construct` a BenchDisk and `prepare` data for read, so I suggest put them in `BenchBuilder::build` and `SimpleDiskBench::prepare` separately.